### ISSUE TITLE
prefer-for-of false positive when index is used to modify array

### ISSUE THREAD

**Issue by buu700** [palantir/tslint#1807]
### Bug Report

- __TSLint version__: `4.0.2`
- __TypeScript version__: `2.0.10`
- __Running TSLint via__: (please choose one) CLI

#### TypeScript code being linted

```ts
const arr = [1, 2, 3];
for (let i = 0 ; i < arr.length ; ++i) {
    arr[i] = 8;
}
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "prefer-for-of": true
    }
}
```

#### Actual behavior

`Expected a 'for-of' loop instead of a 'for' loop with this simple iteration`

#### Expected behavior

No errors, since `i` is being used in a way that wouldn't be possible in a for-of loop.
